### PR TITLE
Prune

### PR DESCRIPTION
**Pull request by pivie**
PRUNE should be usable now. Only tar.gz environments are currently supported, but both local and Work Queue frameworks work.

### CODE REVIEW

**Comment by pivie** [comment]
Ooops, still learning about pull requests. Stick with the 5219e92 commit for this. The next one was just a checkpoint for the next version. I suppose I should make another branch.

**Comment by dthain** [comment]
- Add a Makefile and update configure so that prune is built and installed by default.
- Upgrade readme.txt into doc/prune.html
- Tename the *.plan files to be *.prune, so it's clear they belong to prune.
- The HEP example is nice, but improve the names to be something like simulate, analyze, plot, instead of cms1, cms2, cms3.  And test.plan should be cmstest.prune
- Add a HELP command to prune that briefly summarizes the commands.
- Add commands EXIT, QUIT, and Control-D so that it is possible to quit prune.
-  Configuration of prune-space should be done via command-line arguments, so that they are cleanly separated from the input files.
- Try setting up your example prune scripts like this, so they can be run directly as executables:

```
#!/usr/bin/env prune
PUT cmssw_redhat6.json AS cms_env ENV umbrella
PUT CMS.1 AS cms1
res1 = cms1(10)
...
```

**Comment by dthain** [comment]
The cms example doesn't work for me:
```
dthain@cclws15~/work/cctools/prune/test/hep$ ../../src/prune 
Enter location for data files [/tmp/prune/data/]: 
Enter filepath for the database [/tmp/prune/___prune.db]: 
Enter location for execution sandboxes [/tmp/prune/sandbox/]: 
Warning: No Environments defined!
PRUNE$ PUT cmssw_redhat6.json AS cms_env ENV umbrella
cmssw_redhat6.json 34c659e0-1633-431c-93cc-6d6917c669ab True umbrella  None
PRUNE$ PUT CMS.1 AS cms1                
CMS.1 55c20dc0-a56d-4db6-8ce7-d931e02b50c6 True ascii  None
PRUNE$ res1 = cms1(10)                  
Traceback (most recent call last):
  File "../../src/prune", line 581, in process_line
    res = prunelib.eval(line, 0, cmd_id)
  File "/afs/nd.edu/user37/ccl/work/dthain/cctools/prune/src/prunelib.py", line 410, in eval
    res = eval(remain, depth+1, cmd_id, dict({'assigns':args}.items()+extra.items()) )
  File "/afs/nd.edu/user37/ccl/work/dthain/cctools/prune/src/prunelib.py", line 361, in eval
    op_string += r[3]+','
IndexError: list index out of range
```

**Comment by pivie** [comment]
I should probably remove the CMS example until I get PRUNE working with Umbrella again.

**Comment by dthain** [comment]
If umbrella isn't supported, then a reasonable error message should be given.

**Comment by hmeng-19** [comment]
@pivie , you may want to try the latest version of umbrella.
If your task succeeds, the exit code of umbrella should be 0;
If your task fails, the exit code of umbrella should be 1.

**Comment by pivie** [comment]
Okay, I should have the listed issues resolved in the 0acd4a9 commit. Should I rebase or just leave things as is?

**Comment by dthain** [comment]
- No need to rebase.
- Add basic command line arguments to be consistent with the rest of cctools:  `-d flag|-v|-h`
- The upper-case is easy to read, but annoying to use, please allow lower case.
- Add CAT and EDIT commands so that the user has some easy visibility into prunespace without using GET, which we should discourage.
- After fiddling with it a bit, I now see that keeping track of where the prune config is going to be a pain, as the user navigates around.  So, let's make the default location of the config file `$HOME/.pruneconf` and the default location of prunespace `$HOME/.prunespace`.  Of course, the user can still change it if they want...

- I still get a coredump when running the basic examples.  If I try this:

```
#!/bin/sh

WORKSPACE=/tmp/testbug
INSTALLDIR=${WORKSPACE}/install

rm -r $WORKSPACE
mkdir $WORKSPACE
cd $WORKSPACE

git clone https://github.com/pivie/cctools.git
cd cctools
git checkout prune
./configure --prefix $INSTALLDIR
make
make install

export PATH=$INSTALLDIR/bin:$PATH
export PYTHONPATH=$INSTALLDIR/lib/python2.6/site-packages

cd prune/examples/merge_sort
prune < sort.prune
```

Then I get this:

```
Using default configuration file: /tmp/testbug/cctools/prune/examples/merge_sort/.prune.conf
Use the command line argument "--conf <pathname>" to specify a different one.
Warning: No Environments defined!
PRUNE$ sort.sh 5875422d-700b-48ea-a2bb-a2722a3042a7 True ascii  None
PRUNE$ merge.sh aa432471-86a0-448c-9adf-ef8f61e06780 True ascii  None
PRUNE$ nouns.txt 4db181a0-9ad4-415b-9a67-1e1c21129d35 True ascii  None
PRUNE$ verbs.txt 3ae6a0b8-bad9-4cc3-8741-7d8424ba45a4 True ascii  None
PRUNE$ Allocating 1 local workers. Don't forget to enter the WORK command to start execution.
PRUNE$ {'name': 'sort', 'literal': None, 'io_type': 'F', 'pos': 0, 'file_puid': UUID('5875422d-700b-48ea-a2bb-a2722a3042a7'), 'op_id': '5'}
Traceback (most recent call last):
  File "/tmp/testbug/install/lib/python2.6/site-packages/prune/main.py", line 292, in process_line
    lib.local_check()
  File "/tmp/testbug/install/lib/python2.6/site-packages/prune/lib.py", line 802, in local_check
    operation = create_operation(run['op_id'],'local',local_local_fs)
  File "/tmp/testbug/install/lib/python2.6/site-packages/prune/lib.py", line 568, in create_operation
    if io['pack']=='gzip':
KeyError: 'pack'
```

**Comment by pivie** [comment]
What is in $HOME/.prunespace? Database? Files? Sandboxes?

**Comment by dthain** [comment]
Well, I would think all of them, right?  Together, those three things constitute prunespace.  The average user isn't going to be poking around in those details, except to simply understand that "prunespace" is a distinct namespace from the regular filesystem.

**Comment by pivie** [comment]
Then what is .pruneconf for?

**Comment by pivie** [comment]
I worry that putting everything in $HOME will start people off on the wrong foot. I have had bad experiences putting sqlite in AFS. And putting, in a hidden directory, a system that is designed to consume all space available to it, seems like it is going to cause problems.
Ah, I just realized that perhaps that is what a separate .pruneconf is for. It is small but always there and can redirect the rest of the stuff elsewhere.

**Comment by dthain** [comment]
When starting up, prune loads a config to find the location of prunespace
and other relevant settings. If not given on the command line, then look in
$HOME/.pruneconf.

When setting up prunespace for the first time, suggest
$HOME/.prunespace/{db|files|sandbox} as the default values in the config
file.  (Obviously, you will chose different values for yourself because you
want to work in /pscratch/pivie/prunespace)


Make sense?

**Comment by dthain** [comment]
Ok, keep in mind that AFS is an uncommon oddity particular to Notre Dame and a few other places.  Not many people use it, so you don't want to design the system around it.  It would be reasonable to put in a little hack such that if $HOME starts with /afs, you suggest /tmp/$USER/prunespace when configuring for the first time.

**Comment by pivie** [comment]
We added the LOCATE command to give visibility into the files. If a user tries to CAT a 100GB file, should PRUNE comply?

Also, I'm not sure how you want EDIT to work. I had envisioned that to edit a user would PUT a file in their namespace, edit it with an editor of their choice and GET it back into PRUNE. Are you envisioning an embedded editor of some sort?

**Comment by pivie** [comment]
And sorry for the seg faults. I have two branches and the line between them got blurred when I tried to get set up to work without the CRC on Friday.

**Comment by dthain** [comment]
Good point about cat on a large file, I will defer to your judgement
there.  Maybe only allow it on small files.

If the user does `get;vi;put"  then prune doesn't have any way of
associating the new file with the old and the history is lost.

But if you implement edit as an intrinsic operation, then prune will know
that the file is derived from another.  And it's easy to implement: copy
the file to a temporary sandbox location, run system("$EDITOR $tempfile"),
then archive the changed file and note that it was derived from the
original.  (Of course, be sure to cleanly abort if the file was not changed
or the editor exited with failure.)

**Comment by pivie** [comment]
Are there specific debug flags you are interested in? My assumption was that I would keep a record of debug information in the database since running a prune file can be non-deterministic from a debug perspective because of memoization.

**Comment by pivie** [comment]
Is the version supposed to be the cctools version or a prune version? If cctools, is there an example (in cctools) of how to access that information from python? I seem to remember having this problem with deltadb, but we rewrote it in c and I don't recall if we figured out a solution for python.

**Comment by pivie** [comment]
I assume that the -h argument would say something about the configuration of prune from the command line, and not about the actual PUT,GET,WORK commands?

**Comment by dthain** [comment]
`-d flag` should just call `debug_flags_set` like all our other programs.

**Comment by dthain** [comment]
Use the cctools version, I'm sure you can figure out a way to get the string from `configure` to the source.

**Comment by dthain** [comment]
`-h` should just list the command line arguments, like all our other programs.

**Comment by pivie** [comment]
Let me know if you get more seg faults or want something else added.

**Comment by dthain** [comment]
You seem to have a lot of build failures, please do a `git pull --rebase origin master`, push again, and then see what else needs to be fixed up.

**Comment by pivie** [comment]
I assume mean where origin is cooperative-computing-lab/cctools. (That is upstream not origin for me)

**Comment by dthain** [comment]
Right, that's what I meant.
Also, make sure you are subscribed to cctools-build, so that you get the autobuild messages...

**Comment by dthain** [comment]
Check out the build logs, you have a bug in configure.

**Comment by dthain** [comment]
Suggest you have a talk with @batrick to get some suggestions on the other build failures.

**Comment by pivie** [comment]
@batrick Any ideas why $(CCTOOLS_PYTHON) in a Makefile resolves to '0' on centos? See autobuild for ae563f8819b785156f11516eb85f55ae5bd4ea23

**Comment by batrick** [comment]
Probably because Python isn't installed... (read the configure output)

**Comment by pivie** [comment]
It finally passed autobuild for the -v argument. It is currently working for my census test cases (so a few bugs interfacing with Work Queue have been fixed) and I am trying it on a full scale run right now.

**Comment by dthain** [comment]
Hooray!

**Comment by pivie** [comment]
Please try it again.

**Comment by pivie** [comment]
I think I have implemented all the fixes we talked about this morning.

**Comment by dthain** [comment]
Taking a look.

**Comment by dthain** [comment]
```
 dthain@cclws15~$ prune
PRUNE$ put /usr/share/dict/words stuff
Bad command: use "PUT <local_namespace_filename> AS <prune_namespace_filename>" 
PRUNE$ put /usr/share/dict/words as words
PRUNE$ ls                               
words	-> 4953699 bytes @ 2015-06-01 12:09:25
merged_result	-> 259 bytes @ 2015-06-01 09:13:08
v	-> 115 bytes @ 2015-06-01 09:13:07
n	-> 144 bytes @ 2015-06-01 09:13:05
verbs	-> 115 bytes @ 2015-06-01 09:13:03
nouns	-> 144 bytes @ 2015-06-01 09:13:03
merge	-> 97 bytes @ 2015-06-01 09:13:03
sort	-> 88 bytes @ 2015-06-01 09:13:02
/etc/hosts	-> 327 bytes @ 2015-06-01 09:11:17
PRUNE$ all = sort ( words )
PRUNE$ work                             
No resources are being used, so work can't be done.
Please try "USE local <#>" to use the specified number of processes.
      -or- "USE wq <master_name>" to start a Work Queue master.
PRUNE$ use local 2                      
Allocating 2 local workers. Don't forget to enter the WORK command to start execution.
PRUNE$ work                             
Traceback (most recent call last):
  File "/tmp/cctools/lib/python2.6/site-packages/prune/main.py", line 381, in process_line
    lib.local_check()
  File "/tmp/cctools/lib/python2.6/site-packages/prune/lib.py", line 908, in local_check
    operation = create_operation(run['op_puid'],'local',local_local_fs)
  File "/tmp/cctools/lib/python2.6/site-packages/prune/lib.py", line 634, in create_operation
    env_type = env['env_type']
TypeError: 'NoneType' object is unsubscriptable
```

**Comment by pivie** [comment]
Just added a few commands to the HELP inside prune.

**Comment by pivie** [comment]
Fixed. This was a database migration issue.

**Comment by dthain** [comment]
Cool, I like how the pending evaluation even stays there until the next session:

```
PRUNE$ ls                               
all 	(pending)
words	-> 4953699 bytes @ 2015-06-01 12:09:25
x 	(pending)
merged_result	-> 259 bytes @ 2015-06-01 09:13:08
v	-> 115 bytes @ 2015-06-01 09:13:07
n	-> 144 bytes @ 2015-06-01 09:13:05
verbs	-> 115 bytes @ 2015-06-01 09:13:03
nouns	-> 144 bytes @ 2015-06-01 09:13:03
merge	-> 97 bytes @ 2015-06-01 09:13:03
sort	-> 88 bytes @ 2015-06-01 09:13:02
/etc/hosts	-> 327 bytes @ 2015-06-01 09:11:17
PRUNE$ work                             
No resources are being used, so work can't be done.
Please try "USE local <#>" to use the specified number of processes.
      -or- "USE wq <master_name>" to start a Work Queue master.
PRUNE$ use local 1                      
Allocating 1 local workers. Don't forget to enter the WORK command to start execution.
PRUNE$ work                             
Start: ./sort v
Returned: 0 (./sort v) 
Start: ./sort words
Returned: 0 (./sort words) 
PRUNE finished executing after: 00m07s
PRUNE$                                  
```

**Comment by dthain** [comment]
Seems like I run into these issue pretty easily.  Please take some time to run through all the commands and try them in every different combination that you can think of, and make sure that they test preconditions so that the emit error messages instead of throwing exceptions:

```
PRUNE$ cut words                        
Traceback (most recent call last):
  File "/tmp/cctools/lib/python2.6/site-packages/prune/main.py", line 465, in process_line
    database.var_unset(r['name'])
AttributeError: 'module' object has no attribute 'var_unset'
```

File: configure [hunk 1]
```diff
@@ -1244,6 +1251,12 @@ else
 	echo "*** not building html or man pages"
 fi
 
+# Inject version into prune python code
+PRUNE_MAIN=prune/prune/main.py
+`sed "s/cctools_version = ['\"].*['\"]/cctools_version = '${VERSION}'/g" ${PRUNE_MAIN} > ${PRUNE_MAIN}2`
+`sed "s/cctools_releasedate = ['\"].*['\"]/cctools_releasedate = '${DATE}'/g" ${PRUNE_MAIN}2 > ${PRUNE_MAIN}`
+`rm -f ${PRUNE_MAIN}2`
+
```
Comment (batrick) [inline configure]: This causes `configure` to change the code and working copy. Please use a `conf.py` or similar file which can be ignored. Also, that file should be generated by Make not configure.

File: .gitignore [hunk 2]
```diff
@@ -7,3 +7,5 @@
 cctools.test.log
 configure.rerun
 config.mk
+.prune.conf
```
Comment (batrick) [inline .gitignore]: Please move `prune.conf` to the prune project sub-directory.

**Comment by pivie** [comment]
@dthain Let me know if you find any more stack traces. However, if you have -d all stack traces are purposefully displayed when there is an error.

**Comment by dthain** [comment]
Picking the defaults at first startup causes an exception:

```
dthain@cclws15/tmp/pivie$ prune

Welcome to Prune!
The default config file will be stored at: /afs/nd.edu/user18/dthain/.pruneconf
Enter a location for data files, database, and sandboxes [/afs/nd.edu/user18/dthain/.prunespace/]: /tmp/prunespace
It appears that the database format might have changed. Unless you just RESET the database, you might need to RESET the data in Prune to proceed.
Prune needs a file that describes the environment that operations should be executed in.
Currently the only option is a .tar.gz file that, when unzipped in a sandbox, contains all the libraries and dependencies needed to run the operations in a Linux environment.
What local file should be used as execution environment? [empty.tar.gz] 
Traceback (most recent call last):
  File "/tmp/cctools/lib/python2.6/site-packages/prune/main.py", line 821, in <module>
    lib.initialize(data_folder, sandbox_prefix, hadoop_data)
  File "/tmp/cctools/lib/python2.6/site-packages/prune/lib.py", line 88, in initialize
    return get_default_environment()
  File "/tmp/cctools/lib/python2.6/site-packages/prune/lib.py", line 585, in get_default_environment
    store_file(env_file, ids['puid'])
TypeError: 'NoneType' object is unsubscriptable

PRUNE$ 
```

**Comment by pivie** [comment]
Remember I said you would have to RESET the database. Use "RESET ALL".
However, I will try to prevent this situation also.

**Comment by pivie** [comment]
I guess it WAS a user interface thing to some degree. The default environment file you specified did not exist in the local directory. I checked if it existed before putting it in the database, but not before trying to copy it to the data directory. After swapping two lines it should now force you to specify an EXISTING file for the environment before proceeding. Just as a note, there is a dummy targz environment file in the example.

**Comment by dthain** [comment]
Version zero, more to come!

**Pull request merged**